src/txParsers.ts

```diff
@@ -241,49 +258,95 @@ export const parseCertificates = createParser(parseArray, parseCertificate, Pars
+const parseNetworkId = createParser(parseUint, ParseErrorReason.INVALID_TX_NETWORK_ID)
 
 export const parseTxBody = (unparsedTxBody: unknown): TransactionBody => {
     validate(isMapWithKeysOfType(unparsedTxBody, isNumber), ParseErrorReason.INVALID_TX_BODY_CBOR)
 
     return {
-        inputs: parseInputs(unparsedTxBody.get(0)),
-        outputs: parseOutputs(unparsedTxBody.get(1)),
-        fee: parseFee(unparsedTxBody.get(2)),
-        ttl: parseOptional(unparsedTxBody.get(3), parseTtl),
-        certificates: parseOptional(unparsedTxBody.get(4), parseCertificates),
-        withdrawals: parseOptional(unparsedTxBody.get(5), parseWithdrawals),
-        update: unparsedTxBody.get(6),
-        metadataHash: parseOptional(unparsedTxBody.get(7), parseMetadataHash),
-        validityIntervalStart: parseOptional(unparsedTxBody.get(8), parseValidityIntervalStart),
-        mint: parseOptional(unparsedTxBody.get(9), parseMint),
+        inputs: parseInputs(unparsedTxBody.get(TransactionBodyKeys.INPUTS)),
+        outputs: parseOutputs(unparsedTxBody.get(TransactionBodyKeys.OUTPUTS)),
+        fee: parseFee(unparsedTxBody.get(TransactionBodyKeys.FEE)),
+        ttl: parseOptional(unparsedTxBody.get(TransactionBodyKeys.TTL), parseTtl),
+        certificates: parseOptional(unparsedTxBody.get(TransactionBodyKeys.CERTIFICATES), parseCertificates),
+        withdrawals: parseOptional(unparsedTxBody.get(TransactionBodyKeys.WITHDRAWALS), parseWithdrawals),
+        update: unparsedTxBody.get(TransactionBodyKeys.UPDATE),
+        metadataHash: parseOptional(unparsedTxBody.get(TransactionBodyKeys.METADATA_HASH), parseMetadataHash),
+        validityIntervalStart: parseOptional(unparsedTxBody.get(TransactionBodyKeys.VALIDITY_INTERVAL_START), parseValidityIntervalStart),
+        mint: parseOptional(unparsedTxBody.get(TransactionBodyKeys.MINT), parseMint),
+        scriptDataHash: parseOptional(unparsedTxBody.get(TransactionBodyKeys.SCRIPT_DATA_HASH), parseScriptDataHash),
+        collaterals: parseOptional(unparsedTxBody.get(TransactionBodyKeys.COLLATERAL_INPUTS), parseCollaterals),
+        requiredSigners: parseOptional(unparsedTxBody.get(TransactionBodyKeys.REQUIRED_SIGNERS), parseRequiredSigners),
+        networkId: parseOptional(unparsedTxBody.get(TransactionBodyKeys.NETWORK_ID), parseNetworkId),
     }
 }
 
 export const parseTx = (unparsedTx: unknown): Transaction => {
-    const [body, witnessSet, auxiliaryData] = parseTuple(
+    const [body, ...otherItems] = parseTuple(
         unparsedTx,
         ParseErrorReason.INVALID_TX_CBOR,
         parseTxBody,
-        dontParse,
-        dontParse,
+        //             | shelley era:  | alonzo era:
+        dontParse,  // | witnessSet    | witnessSet
+        dontParse,  // | auxiliaryData | scriptValidity
+        dontParse,  // | `undefined`   | auxiliaryData
     )
+    const presentItems = otherItems.filter((item) => item !== undefined)
```

Comment:
This looks suspicious. Is it true that we are actually removing only the third (i.e. [2]) item (or nothing)? It would be safer to assert that is true before we remove anything (so that we catch any future changes immediately).

Reply:
The `undefined` values should appear only at the end of `otherItems` (they are created by `parseTuple` when the item count is lower than parser count; cardano-cli seems to use `null` if it wants to communicate an empty value), so validating the item count should be sufficient, right? Added in https://github.com/vacuumlabs/cardano-hw-interop-lib/pull/3/commits/f7b82cf0005d25fabe891e52116525a44f45d9b4.

Reply:
It seems we are relying on third-party stuff, like guessed/implicit behavior of `cardano-cli`, or the generic cbor parsing library. I don't feel comfortable relying on the difference between null/undefined if the value is produced too far from this place (or even not by us).

Let's add some validation/assertion for "The undefined values should appear only at the end of otherItems".

Reply:
Ok, added in https://github.com/vacuumlabs/cardano-hw-interop-lib/pull/3/commits/f85f8f3749915e56a281312aafffbb1113dfbef5